### ISSUE TITLE
New vcpkg PR advice

### ISSUE THREAD

**Issue by yintothayang** [Microsoft/vcpkg#1108]
Hello! 

I have built a new vcpkg port to easily distribute my library, and would like to publish it to vcpkg ports.  I have read the contributing.md and have signed the CLA. I was hoping for some feedback / potential changes I need to make, before I send a PR. Thanks in advance!

The portfile itself is simple:

```
include(vcpkg_common_functions)

vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO Innervate/red-shell-cpp
    REF 1.0.0
    SHA512 97989d8d323ff6a7f43b6328894458e48de8965c412cd13b5ec7e3345834d3d3291dc8127bf12d93b7872dc115fdaede7a4cd7eee53720dcddb4c8df3739894e
    HEAD_REF master
)

set(SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/redshell)

file(MAKE_DIRECTORY
  ${CURRENT_PACKAGES_DIR}/include/redshell
  #${CURRENT_PACKAGES_DIR}/lib
  #${CURRENT_PACKAGES_DIR}/debug/lib  
)

# Header .h
file(COPY
    "${SOURCE_PATH}/include/redshell.h"
    DESTINATION ${CURRENT_PACKAGES_DIR}/include/redshell
)

# Debug .lib
file(COPY
    "${SOURCE_PATH}/debug/lib/redshell.lib"
    DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib
)

# Release .lib
file(COPY
    "${SOURCE_PATH}/lib/redshell.lib"
    DESTINATION ${CURRENT_PACKAGES_DIR}/lib
)

# Copyright
file(WRITE ${CURRENT_PACKAGES_DIR}/share/redshell/copyright "2017 Innervate Inc. All Rights Reserved.")
```

Control:

```
Source: redshell
Build-Depends: cpprestsdk [windows]
Version: 1.0.0
Description: redshell sdk redshell.io
```

**Comment by Mixaill** [comment]
No source code? ☹️ 

>https://red-shell-staging.gamesight.io/pricing

Is it paid software? Don't think that vcpkg project accepts paid software.


---

You should add something like in the begging of the file (but after include)

```
if(NOT VCPKG_TARGET_ARCHITECTURE STREQUAL x86)
  message(FATAL_ERROR "Error: redshell only supports the x86 architecture.")
endif()

if (VCPKG_CMAKE_SYSTEM_NAME STREQUAL WindowsStore)
    message(FATAL_ERROR "Error: UWP builds is not supported.")
endif()

if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
    message(STATUS "Warning: Dynamic building not supported yet. Building static.")
    set(VCPKG_LIBRARY_LINKAGE static)
endif()
```
---
Control should be like 
```
Source: redshell
Version: 1.0.0
Description: Helps Steam games uncover where their players come from through reliable attribution
Build-Depends: cpprestsdk
```

**Comment by ras0219-msft** [comment]
@yintothayang Thanks for the interest in Vcpkg! We would like to have your library, if we can find a way to make it meet the needs of Vcpkg.

We are willing to accept non-free software that plays well with the rest of our ecosystem and doesn't compromise our core user experience. In this case, it looks like the (important?) non-free/business component is actually the backing REST API, so I hope this means that the red-shell-cpp _client_ code can actually be open-sourced.

We _can_ accept prebuilt libraries, but they must not externally depend on any other libraries. This is because it would restrict the ability for those dependencies to be updated. Because this library depends on `cpprestsdk` in its header file (not just as an implementation detail), I don't see a way for that to be accomplished without publishing the source code.

To put it simply: what's your plan to handle ABI-incompatible updates in your dependencies?

Additionally, while I Am Not A Lawyer, I believe we will need more of a copyright file than "2017 Innervate Inc. All Rights Reserved.". I see no way for general users to use the library given no license, so I wouldn't be comfortable adding it to the tree.

Even though all the above points are negative, I really would like to add your library if we can get things sorted out!

The most straightforward way to solve all the above problems would be to publish your _client_ code under an MIT/BSD/Apache license (your REST service is totally fine as-is). We could rebuild from source, which would resolve our concerns about cpprestsdk updates and it will also allow us to apply patches to work around any cpprestsdk API breaking changes.

As an example of this exact pattern -- open source client code backed by a non-free service -- see `azure-storage-cpp` (which uses cpprestsdk!) and `aws-sdk-cpp`.

**Comment by yintothayang** [comment]
@ras0219-msft Thanks for the quick feedback! 
@Mixaill  Thanks for the suggestions on my CONTROL and portfile!

I'm quite new to native c++ windows, I took your suggestions and think I've come up with a solution that meets both of our needs. Please correct me if i'm wrong, you had 2 main concerns I tried to correct with my latest build.  

1. License Concern 

The original license was just a placeholder, sorry about that. I have since updated to a [MIT license](https://github.com/Innervate/red-shell-cpp/blob/master/LICENSE.txt), which will be copied via the new portfile (ref at the bottom)

2. ABI-incompatible updates in dependencies 

I have refactored the code, so the  [header](https://github.com/Innervate/red-shell-cpp/blob/master/include/RedShell.h) no longer takes dependency on `cpprestsdk`.  The library implementation has very minimal interaction with `cpprestsdk`. 

Unfortunately, releasing our source wouldn't make sense for our business at this time. I hope, and believe that the changes I've made will be satisfactory in decoupling our dependency on `cpprestsdk`. 

Looking forward to your feedback, Thank you!


### Updated port files
> CONTROL

```
Source: redshell
Build-Depends: cpprestsdk [windows]
Version: 1.0.0
Description: RedShell SDK, helping game devs attribute Steam installs -> https://github.com/Innervate/red-shell-cpp
```

> portfile.cmake 
```
include(vcpkg_common_functions)

set(SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/redshell)

vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO Innervate/red-shell-cpp
    REF 1.0.1
    SHA512 e851a9b5fbcc952af92d5fc4dfd1f23500cb6afa77bfbb7055c2120d226c371f5dea8c703ce531b5128e5e4b85397fe6bf3c67387a31faab48372778c688ac36
    HEAD_REF master
)

#file(MAKE_DIRECTORY
  #${CURRENT_PACKAGES_DIR}/include/redshell
  #${CURRENT_PACKAGES_DIR}/lib
  #${CURRENT_PACKAGES_DIR}/debug/lib  
####)

# Header .h
file(COPY
    "${SOURCE_PATH}/include/RedShell.h"
    DESTINATION ${CURRENT_PACKAGES_DIR}/include/redshell
)

# Debug .lib
file(COPY
    "${SOURCE_PATH}/lib/${VCPKG_TARGET_ARCHITECTURE}/debug/RedShell.lib"
    DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib
)

# Release .lib
file(COPY
    "${SOURCE_PATH}/lib/${VCPKG_TARGET_ARCHITECTURE}/release/RedShell.lib"
    DESTINATION ${CURRENT_PACKAGES_DIR}/lib
)

# Copyright
file(COPY
    "${SOURCE_PATH}/LICENSE.txt"
    DESTINATION ${CURRENT_PACKAGES_DIR}/share/redshell/copyright
)
```

**Comment by ras0219-msft** [comment]
Thanks for making these changes.

Unfortunately, to fully decouple from cpprestsdk, you'll also need to statically link cpprestsdk (and any further dependencies) into your DLL (you'll need to produce a DLL instead of a static library). The objective is to enable users to update the cpprestsdk package -- which could produce a _completely incompatible output_ -- without breaking your library.

I know this seems like a lot of change, but this is actually a problem you'll hit with or without Vcpkg. Let me explain with an example that doesn't involve Vcpkg at all:

---

Developer Bob downloads your prebuilt library and the current version of cpprestsdk. They're really happy with your service, build an application, and ship it. Everything's great!

Now, they want to add some additional networking to their app and want to use cpprestsdk. They want to use some features from (let's say) a new version 3.0. They rebuild cpprestsdk from the latest GitHub, which now builds a `cpprest_3_0.dll` binary and there's no longer a `cpprest_2_9.dll`. Furthermore, all of the string APIs were switched from UTF16 to UTF8 and are completely binary incompatible with the previous ones! RedShell then fails to load because of the ABI break and there's no way Bob can fix the problem on his own 😢. Bob now has to choose what he wants more: the new cpprestsdk APIs/bugfixes/security or the current functionality in redshell-cpp.

We don't want to make developers choose and I'm sure you don't either -- imagine if they decided redshell wasn't important enough to block their other upgrade! This is important not just for new features, but also for security.

_Note: in this example, a solution may exist by shipping both cpprestsdk versions, however this is not always possible._

---

#### How to solve this

The solution to this problem (while still having a prebuilt library) is that RedShell must:
- ✔️ Completely hide uses of other (non-abi-stable) libraries in headers
- ✔️ Link against the dynamic CRT
- ❌ Completely hide references to other libraries from any binaries
- ❌ [Optional, enables compiler upgrades] Use only C on the DLL boundary.

To accomplish "Completely hide references to other libraries from your binary", you must make a DLL which static links all your non-CRT dependencies. Static libs still need to be linked, which depends on the actual symbol names of the library you're trying to link against -- if they change, you'll fail at link time.

Vcpkg can help you accomplish this. You will need to create a new file `triplets\x86-windows-static-md.cmake` with the contents:
```cmake
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
```

This will let you use `vcpkg install cpprestsdk:x86-windows-static-md` to build a statically linked version of cpprestsdk that uses the dynamic CRT runtime. You'll then need to follow the instructions on our [static linking blog post](https://blogs.msdn.microsoft.com/vcblog/2016/11/01/vcpkg-updates-static-linking-is-now-available/) to set your project (RedShell) to use the new `x86-windows-static-md` triplet.

For MSBuild, you will need something like this:
```xml
<PropertyGroup Label="Globals">
    ....
    <VcpkgTriplet Condition="'$(Platform)'=='Win32'">x86-windows-static-md</VcpkgTriplet>
    <VcpkgTriplet Condition="'$(Platform)'=='x64'">x64-windows-static-md</VcpkgTriplet>
    <VcpkgEnabled>true</VcpkgEnabled>
</PropertyGroup>
```

Then, your output binary should no longer depend on `cpprest_2_9.dll`. You can confirm this by opening a Visual Studio Developer Prompt and running:
```
dumpbin /dependents RedShell.dll
```

You can repeat the above steps again for 64-bit by creating a `triplets\x64-windows-static-md.cmake` that sets `VCPKG_TARGET_ARCHITECTURE` to `x64`.

I'd be happy to help further over email at vcpkg@microsoft.com.

**Comment by yintothayang** [comment]
Email sent @ras0219-msft , thanks for all the help.

**Comment by yintothayang** [comment]
I have followed all of the steps listed, vcpkg built the `x86-windows-static-md` packages just fine. I updated my property pages to add those Global `VcpkgTriplet` variables.  The Headers show up in my Visual Studio project, and everything seems to be linked correctly.  

Unfortunately I'm now stuck on [this issue](https://github.com/Microsoft/cpprestsdk/issues/344) with cpprestsdk preprocessor definitions.
